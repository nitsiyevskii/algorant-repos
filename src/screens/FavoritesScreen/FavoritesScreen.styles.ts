import { createStyles } from "src/utils/createStyles";

export const useStyles = createStyles((theme) => ({
  safeArea: {
    flex: 1,
    backgroundColor: theme.colors.background.primary,
  },
  container: {
    flex: 1,
  },
  listContainer: {
    padding: 16,
    flexGrow: 1,
  },
}));

