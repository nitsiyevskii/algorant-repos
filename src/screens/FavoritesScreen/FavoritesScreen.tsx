import React from "react";
import { View, FlatList } from "react-native";
import { EmptyState } from "src/components/EmptyState/EmptyState";
import { RepositoryCard } from "src/components/RepositoryCard/RepositoryCard";
import { GitHubRepository } from "src/types";
import { useAppDispatch, useAppSelector } from "src/store/hooks";
import { toggleFavorite } from "src/store/slices/repositoriesSlice";
import {
  selectFavoriteRepositories,
  selectFavorites,
} from "src/store/selectors/repositoriesSelectors";
import { useStyles } from "./FavoritesScreen.styles";

const FavoritesScreen = () => {
  const styles = useStyles();
  const dispatch = useAppDispatch();

  const favoriteRepositories = useAppSelector(selectFavoriteRepositories);
  const favorites = useAppSelector(selectFavorites);

  const handleToggleFavorite = (id: number) => {
    dispatch(toggleFavorite(id));
  };

  const isFavorite = (id: number) => favorites.includes(id);

  const renderRepository = ({ item }: { item: GitHubRepository }) => (
    <RepositoryCard
      repository={item}
      isFavorite={isFavorite(item.id)}
      onToggleFavorite={() => handleToggleFavorite(item.id)}
    />
  );

  const renderEmpty = () => (
    <EmptyState
      icon="heart-outline"
      message="No favorites yet"
      description="Star repositories to see them here"
    />
  );

  return (
    <View style={styles.container}>
      <FlatList
        data={favoriteRepositories}
        renderItem={renderRepository}
        keyExtractor={(item) => item.id.toString()}
        contentContainerStyle={styles.listContainer}
        ListEmptyComponent={renderEmpty}
      />
    </View>
  );
};

export default FavoritesScreen;
