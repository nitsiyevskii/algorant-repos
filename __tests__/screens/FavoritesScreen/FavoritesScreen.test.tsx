import React from 'react';
import { render } from '@testing-library/react-native';
import { useDispatch, useSelector } from 'react-redux';
import FavoritesScreen from 'src/screens/FavoritesScreen/FavoritesScreen';

jest.mock('react-redux');

const mockDispatch = jest.fn();
const mockUseDispatch = useDispatch as unknown as jest.Mock;
const mockUseSelector = useSelector as unknown as jest.Mock;

const mockRepositories = [
  {
    id: 1,
    name: 'favorite-repo',
    full_name: 'algorand/favorite-repo',
    owner: {
      login: 'algorand',
      avatar_url: 'https://test.com/avatar.png',
      html_url: 'https://github.com/algorand',
    },
    html_url: 'https://github.com/algorand/favorite-repo',
    description: 'Favorite repository',
    fork: false,
    created_at: '2023-01-01T00:00:00Z',
    updated_at: '2023-01-02T00:00:00Z',
    pushed_at: '2023-01-02T00:00:00Z',
    stargazers_count: 200,
    watchers_count: 100,
    language: 'JavaScript',
    forks_count: 20,
    open_issues_count: 10,
    topics: ['blockchain'],
    visibility: 'public',
    default_branch: 'main',
  },
];

describe('FavoritesScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockUseDispatch.mockReturnValue(mockDispatch);
    
    mockUseSelector.mockImplementation((selector) => {
      const state = {
        repositories: {
          items: mockRepositories,
          favorites: [1],
          isLoading: false,
          error: null,
        },
        filters: {
          organizations: [],
          languages: [],
          stars: { min: null, max: null },
          forks: { min: null, max: null },
          searchQuery: '',
        },
      };
      return selector(state);
    });
  });

  it('renders correctly', () => {
    const { getByText } = render(<FavoritesScreen />);
    expect(getByText('favorite-repo')).toBeTruthy();
  });

  it('displays favorite repositories', () => {
    const { getByText } = render(<FavoritesScreen />);
    
    expect(getByText('favorite-repo')).toBeTruthy();
    expect(getByText('algorand')).toBeTruthy();
    expect(getByText('Favorite repository')).toBeTruthy();
  });

  it('shows empty state when no favorites', () => {
    mockUseSelector.mockImplementation((selector) => {
      const state = {
        repositories: {
          items: [],
          favorites: [],
          isLoading: false,
          error: null,
        },
        filters: {
          organizations: [],
          languages: [],
          stars: { min: null, max: null },
          forks: { min: null, max: null },
          searchQuery: '',
        },
      };
      return selector(state);
    });

    const { getByText } = render(<FavoritesScreen />);
    
    expect(getByText('No favorites yet')).toBeTruthy();
    expect(getByText('Star repositories to see them here')).toBeTruthy();
  });

  it('displays repository statistics', () => {
    const { getByText } = render(<FavoritesScreen />);
    
    expect(getByText('200')).toBeTruthy();
    expect(getByText('JavaScript')).toBeTruthy();
    expect(getByText('20')).toBeTruthy();
  });

  it('renders favorite button as selected for favorite repos', () => {
    const { getByTestId } = render(<FavoritesScreen />);
    const favoriteButton = getByTestId('favorite-button');
    
    expect(favoriteButton).toBeTruthy();
  });

  it('matches snapshot with favorites', () => {
    const { toJSON } = render(<FavoritesScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot when empty', () => {
    mockUseSelector.mockImplementation((selector) => {
      const state = {
        repositories: {
          items: [],
          favorites: [],
          isLoading: false,
          error: null,
        },
        filters: {
          organizations: [],
          languages: [],
          stars: { min: null, max: null },
          forks: { min: null, max: null },
          searchQuery: '',
        },
      };
      return selector(state);
    });

    const { toJSON } = render(<FavoritesScreen />);
    expect(toJSON()).toMatchSnapshot();
  });
});

