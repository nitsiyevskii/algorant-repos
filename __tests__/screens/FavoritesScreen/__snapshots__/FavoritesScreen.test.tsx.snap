// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesScreen matches snapshot when empty 1`] = `
<View
  edges={
    [
      "top",
    ]
  }
  style={
    {
      "backgroundColor": "#F2F2F7",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={[Function]}
      contentContainerStyle={
        {
          "flexGrow": 1,
          "padding": 16,
        }
      }
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "paddingHorizontal": 32,
            }
          }
        >
          <Text
            color="#8E8E93"
            name="heart-outline"
            size={64}
            testID="icon"
          >
            heart-outline
          </Text>
          <Text
            style={
              {
                "color": "#3C3C43",
                "fontSize": 18,
                "fontWeight": "500",
                "marginTop": 16,
                "textAlign": "center",
              }
            }
          >
            No favorites yet
          </Text>
          <Text
            style={
              {
                "color": "#8E8E93",
                "fontSize": 16,
                "marginTop": 8,
                "textAlign": "center",
              }
            }
          >
            Star repositories to see them here
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`FavoritesScreen matches snapshot with favorites 1`] = `
<View
  edges={
    [
      "top",
    ]
  }
  style={
    {
      "backgroundColor": "#F2F2F7",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={[Function]}
      contentContainerStyle={
        {
          "flexGrow": 1,
          "padding": 16,
        }
      }
      data={
        [
          {
            "created_at": "2023-01-01T00:00:00Z",
            "default_branch": "main",
            "description": "Favorite repository",
            "fork": false,
            "forks_count": 20,
            "full_name": "algorand/favorite-repo",
            "html_url": "https://github.com/algorand/favorite-repo",
            "id": 1,
            "language": "JavaScript",
            "name": "favorite-repo",
            "open_issues_count": 10,
            "owner": {
              "avatar_url": "https://test.com/avatar.png",
              "html_url": "https://github.com/algorand",
              "login": "algorand",
            },
            "pushed_at": "2023-01-02T00:00:00Z",
            "stargazers_count": 200,
            "topics": [
              "blockchain",
            ],
            "updated_at": "2023-01-02T00:00:00Z",
            "visibility": "public",
            "watchers_count": 100,
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#FFFFFF",
                "borderColor": "#C6C6C8",
                "borderRadius": 12,
                "borderWidth": 1,
                "marginBottom": 12,
                "opacity": 1,
                "padding": 16,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "flex-start",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginBottom": 8,
                }
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                    "marginRight": 12,
                  }
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    {
                      "color": "#000000",
                      "fontSize": 18,
                      "fontWeight": "600",
                      "marginBottom": 4,
                    }
                  }
                >
                  favorite-repo
                </Text>
                <Text
                  numberOfLines={1}
                  style={
                    {
                      "color": "#3C3C43",
                      "fontSize": 14,
                    }
                  }
                >
                  algorand
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": false,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#F2F2F7",
                    "borderRadius": 20,
                    "height": 40,
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": 40,
                  }
                }
                testID="favorite-button"
              >
                <Text
                  color="#FF3B30"
                  name="heart"
                  size={24}
                  testID="icon"
                >
                  heart
                </Text>
              </View>
            </View>
            <Text
              numberOfLines={2}
              style={
                {
                  "color": "#3C3C43",
                  "fontSize": 16,
                  "lineHeight": 20,
                  "marginBottom": 12,
                }
              }
            >
              Favorite repository
            </Text>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 16,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 4,
                  }
                }
              >
                <Text
                  color="#8E8E93"
                  name="star"
                  size={16}
                  testID="icon"
                >
                  star
                </Text>
                <Text
                  style={
                    {
                      "color": "#8E8E93",
                      "fontSize": 14,
                    }
                  }
                >
                  200
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 4,
                  }
                }
              >
                <View
                  style={
                    [
                      {
                        "borderRadius": 6,
                        "height": 12,
                        "width": 12,
                      },
                      {
                        "backgroundColor": "#007AFF",
                      },
                    ]
                  }
                />
                <Text
                  style={
                    {
                      "color": "#8E8E93",
                      "fontSize": 14,
                    }
                  }
                >
                  JavaScript
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 4,
                  }
                }
              >
                <Text
                  color="#8E8E93"
                  name="git-branch"
                  size={16}
                  testID="icon"
                >
                  git-branch
                </Text>
                <Text
                  style={
                    {
                      "color": "#8E8E93",
                      "fontSize": 14,
                    }
                  }
                >
                  20
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
